R/read-write.R

```diff
@@ -0,0 +1,428 @@
+    if (overwrite) {
+      message(glue::glue("Overwriting file {tick(xlsx_file_path)}."))
+    } else{
+      abort_file_exists(xlsx_file_path)
+    }
+  }
+
+  xlsx_tables <- prepare_tables(dm)
+
+  xl_sheet_list <- c(
+    dm_get_tables_impl(dm),
+    list(
+      "___info_dm" = xlsx_tables$info_tibble,
+      "___coltypes_dm" = xlsx_tables$col_class_tibble,
+      "___pk_dm" = xlsx_tables$pk_tibble,
+      "___fk_dm" = xlsx_tables$fk_tibble
+      )
+    )
+
+  writexl::write_xlsx(xl_sheet_list, path = xlsx_file_path)
+  message(
+    glue::glue("Written `dm` as xlsx-file {tick(xlsx_file_path)}.")
+  )
+  invisible(xlsx_file_path)
+}
+
+#' @details `dm_read_xlsx()`: read a `dm` from an `xlsx`-file created using `dm_write_xlsx()`.
+#' @rdname dm-read-write
+#' @export
+dm_read_xlsx <- function(xlsx_file_path) {
+  sheet_list <- readxl::excel_sheets(xlsx_file_path)
+  special_sheets <- c(
+    "___info_dm",
+    "___coltypes_dm",
+    "___pk_dm",
+    "___fk_dm"
+  )
+  if (!all(special_sheets %in% sheet_list)) {
+    abort_files_or_sheets_missing(
+      setdiff(special_sheets, intersect(sheet_list[1:3], special_sheets)),
+      xlsx_file_path,
+      csv = FALSE
+    )
+  }
+
+  def_base <- readxl::read_xlsx(
+    xlsx_file_path,
+    sheet = "___info_dm",
+    col_types = "text"
+  )
+  table_names <- def_base$table
+
+  col_class_info <- readxl::read_xlsx(
+    xlsx_file_path,
+    sheet = "___coltypes_dm",
+    col_types = "text"
+  ) %>%
+    group_by(table) %>%
+    summarize(column = list(column), class = list(class))
+  # FIXME: use it wisely
+
+  pk_info <- readxl::read_xlsx(
+      xlsx_file_path,
+      sheet = "___pk_dm",
+    col_types = "text"
+  )
+
+  fk_info <- readxl::read_xlsx(
+    xlsx_file_path,
+    sheet = "___fk_dm",
+    col_types = c("text", "text", "text", "numeric")
+  )
+
+  table_sheets <- setdiff(
+    sheet_list,
+    c("___info_dm", "___pk_dm", "___fk_dm", "___coltypes_dm")
+  )
+  # sort table_files according to column def_base$table
+  table_sheets_sorted <- table_sheets[order(match(table_sheets, table_names))]
+
+  table_tibble <- map(
+    table_sheets_sorted,
+    readxl::read_xlsx, path = xlsx_file_path
+  ) %>%
+    set_names(table_names) %>%
+    enframe(name = "table", value = "data")
+
+  # since reading from xlsx via `readxl::read_xlsx()` does e.g. not support reading as integer
+  # we reclass the columns in retrospect:
+  table_tibble_reclassed <- left_join(table_tibble, col_class_info, by = "table") %>%
+    select(-table) %>%
+    pmap(function(data, column, class) {
+      reduce2(column, class, function(data, column, class) {
+      mutate(data, !!column := get(paste0("as.", class))(!!sym(column)))
+    }, .init = data)
+    }) %>%
+    set_names(table_names) %>%
+    enframe(name = "table", value = "data")
+
+  make_dm(def_base, table_names, table_tibble_reclassed, pk_info, fk_info)
+}
+
+make_dm <- function(def_base, table_names, table_tibble, pk_info, fk_info) {
```

Comment:
This looks similar to the older `new_dm*()` functions, any overlap?

Reply:
I had a bet with myself, that you would wonder about that. 
I did wonder, too, what's the best way to recreate the `dm`, but just came up with two options:
1. the one that I chose
2. using `dm()` on the list of tables (after repairing the classes) and then using the metainfo to set the key-constraints on the raw `dm`.

On a different note: I am quite sure, that the current implementation can't work with compound keys. E.g., how could the foreign key column know, which primary key column it should refer to? 
Maybe I should have a look into the compound key PR first.

Reply:
If we still allow only one primary key per table? I think it might just work.

The implementation is fine, let's make sure there's no overlap with existing code.

Reply:
ok, nice.

I didn't see overlap with existing code.